README.md

```diff
@@ -1,28 +1,273 @@
 # Muffin
 
-Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/muffin`. To experiment with that code, run `bin/console` for an interactive prompt.
+## Why form objects?
 
-TODO: Delete this and the text above, and describe your gem
+Form objects encapsulate logic to modify data (similar to changesets in Elixir or Mutations in GraphQL). Every non trivial form in rails usually has some custom (and conditional) validation, specific behavior (when update `x`, then remove `y`) and complex association (e.g. `accepts_nested_attributes_for`). This is usually spread all over the model leading to hard to maintain code and tons of conditional validation that is hard to understand. Also it’s not possible to have a form for many objects without having a parent object that is doing the nested_attribute dance.
 
-## Installation
+Forms are living in `/app/forms`. They should work independently from controllers (for unit testing) and can (but doesn’t have to!) handle `ActiveRecord` objects.
 
-Add this line to your application's Gemfile:
+## Public API
 
 ```ruby
-gem 'muffin'
+my_form = MyForm.new(request:, params:, scope:) # scope could be the current_user
+my_form.call # 'commits' the form: it validates and calls the internal process method. returns true on success, false when validation fails. Other errors are signaled via Exceptions.
+my_form.call! # same as call, but raises a ValidationError if validation fails
 ```
 
-And then execute:
+## Attributes
 
-    $ bundle
+Attributes specify which attribute in the form can be set via a request
 
-Or install it yourself as:
+```ruby
+class MyForm < Muffin::Base
+  attribute :name # type String is implicit
+  attribute :age, Integer # second argument defines type if present
+  attribute :accepted?, Boolean # boolean is defined for true or false, converts strings like "on" or "off" (from forms) automatically to their boolean value
+  attribute :tags, array: true # array of strings
+  attribute :tags, [String] # same as above
+end
+```
+
+Forms can contain validations
+
+```ruby
+class MyForm < Muffin::Base
+  attribute :name
+  validates :name, presence: true
+end
+```
+
+And also give a list of required attributes (useful for html validation and marking them in the UI).
+
+```ruby
+my_form.required_attributes # [:name]
+my_form.valid? # returns true/false
+my_form.errors # returns an error object
+```
+
+Attributes are automatically assigned on init:
+
+my_form = MyForm.new(params: { name: "Superman" }) # assigns the name attribute
+my_form.attributes # { name: "Superman" }
+
+## Performing Changes
+
+When `call` is invoked, the form performs validation steps. If those steps are successful, `perform` is called. `perform` is invoked inside of a transaction. If any invalidity errors (e.g. from `save!`) are raised, this will take care of it and update the errors object accordingly.
+
```

Comment:
actually that's not the case anymore. `perform!` will reraise all errors, including validation errors.